harness/determined/exec/tensorboard.py

```diff
@@ -26,121 +40,119 @@ def set_s3_region() -> None:
+    if version.parse(tb_version) >= version.parse("2.5"):
+        tensorboard_args.append("--load_fast=false")
 
-    tensorboard_args.append(f"--logdir={logdir}")
-    return tensorboard_args
+    tensorboard_args.append(f"--logdir={tfevents_dir}")
 
+    return tensorboard_args
 
-def main(args: List[str]) -> int:
-    set_s3_region()
 
+def get_tensorboard_url() -> str:
     task_id = os.environ["DET_TASK_ID"]
     port = os.environ["TENSORBOARD_PORT"]
     tensorboard_addr = f"http://localhost:{port}/proxy/{task_id}"
-    url = f"{tensorboard_addr}/data/plugin/scalars/tags"
-    tensorboard_args = get_tensorboard_args(args)
+    return tensorboard_addr
+
+
+def check_tensorboard_responsive() -> bool:
+    # Ensure Tensorboard is responding to HTTP request to prevent 502 from master.
+    tensorboard_url = get_tensorboard_url()
+    try:
+        # Attempt HTTP request to Tensorboard.
+        res = requests.get(tensorboard_url)
+        res.raise_for_status()
+        return True
+
+    except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError, ValueError) as exp:
+        logger.warning(f"Tensorboard not responding to HTTP: {exp}")
+
+    return False
+
 
-    print(f"Running: {tensorboard_args}")
-    p = subprocess.Popen(tensorboard_args)
+def start_tensorboard(
+    config: Dict[str, Any],
+    tb_version: str,
+    storage_paths: List[str],
+    add_tb_args: List[str],
+) -> int:
+    """Start Tensorboard and look for new files."""
 
-    def still_alive() -> bool:
-        return p.poll() is None
+    stop_time = time.time() + MAX_WAIT_TIME
+    triggered = False
+    responsive = False
 
-    if not wait_for_tensorboard(600, url, still_alive):
-        p.kill()
+    with tempfile.TemporaryDirectory() as local_dir:
 
-    return p.wait()
+        # Get fetcher and perform initial fetch
+        fetcher = fetchers.build(config, storage_paths, local_dir)
+        num_fetched_files = fetcher.fetch_new()
+
+        # Build Tensorboard args and launch process.
+        tb_args = get_tensorboard_args(tb_version, local_dir, add_tb_args)
+        logger.debug(f"tensorboard args: {tb_args}")
+        tensorboard_process = subprocess.Popen(tb_args)
+
+        try:
+            while True:
+                ret_code = tensorboard_process.poll()
+                if ret_code is not None:
+                    raise RuntimeError(f"Tensorboard process died, exit code({ret_code}).")
+
+                # Check if we have reached a timeout without receiving metrics
+                if num_fetched_files == 0 and time.time() > stop_time:
+                    raise RuntimeError("Tensorboard wasn't responsive before the timeout.")
+
+                if responsive and not triggered and num_fetched_files > 0:
+                    print(TENSORBOARD_TRIGGER_READY_MSG)
+                    triggered = True
+
+                time.sleep(FETCH_INTERVAL)
+                responsive = check_tensorboard_responsive()
```

Comment:
you probably need `if not responsive: ...` here.  No point in checking tensorboard over and over I think.

Reply:
Ahh, yeah, I reworked this and that slipped through... it was there before somewhere.